diff --git a/init/main.c b/init/main.c
index 02f6b6bd6a17..2da83da8a9ba 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1108,9 +1108,17 @@ static int __ref kernel_init(void *unused)
 
 	system_state = SYSTEM_RUNNING;
 	numa_default_policy();
+	/*
+	 *
+	 * TROST_JONATHAN: Modificaiton for P0.  
+	 *
+	 */
+
+	printk(KERN_ERR "\n##### Jonathan Trost (UFID: 6736-9531) Welcome to reptilian, there are many like it, but this one is mine. #####\n\n");
 
 	rcu_end_inkernel_boot();
 
+
 	if (ramdisk_execute_command) {
 		ret = run_init_process(ramdisk_execute_command);
 		if (!ret)
@@ -1138,8 +1146,10 @@ static int __ref kernel_init(void *unused)
 	    !try_to_run_init_process("/bin/sh"))
 		return 0;
 
+
 	panic("No working init found.  Try passing init= option to kernel. "
 	      "See Linux Documentation/admin-guide/init.rst for guidance.");
+
 }
 
 static noinline void __init kernel_init_freeable(void)
